shared/domains/chelonia/GIMessage.js

```diff
@@ -38,59 +53,97 @@ export class GIMessage {
   static OP_PROTOCOL_UPGRADE: 'pu' = 'pu'
   static OP_PROP_SET: 'ps' = 'ps' // set a public key/value pair
   static OP_PROP_DEL: 'pd' = 'pd' // delete a public key/value pair
+  static OP_CONTRACT_AUTH: 'ca' = 'ca' // authorize a contract
+  static OP_CONTRACT_DEAUTH: 'cd' = 'cd' // deauthorize a contract
+  static OP_ATOMIC: 'at' = 'at' // atomic op
+  static OP_KEYSHARE: 'ks' = 'ks' // key share
+  static OP_KEY_REQUEST: 'kr' = 'kr' // key request
+  static OP_KEY_REQUEST_RESPONSE: 'krr' = 'krr' // key request response
 
   // eslint-disable-next-line camelcase
   static createV1_0 (
-    contractID: string | null = null,
-    previousHEAD: string | null = null,
-    op: GIOp,
-    manifest: string,
-    signatureFn?: Function = defaultSignatureFn
+    {
+      contractID,
+      originatingContractID,
+      previousHEAD = null,
+      op,
+      manifest,
+      signatureFn = defaultSignatureFn
+    }: {
+      contractID: string | null,
+      originatingContractID?: string,
+      previousHEAD?: string | null,
+      op: GIOp,
+      manifest: string,
+      signatureFn?: Function
+    }
   ): this {
-    const message = {
+    const head = {
       version: '1.0.0',
       previousHEAD,
       contractID,
-      op,
-      manifest,
-      // the nonce makes it difficult to predict message contents
-      // and makes it easier to prevent conflicts during development
-      nonce: Math.random()
```

Comment:
Why did you get rid of the `nonce`?

Reply:
Because: (1) all signed messages (which should be all messages) should already be unique due to the randomness in the signature, (2) the nonce doesn't really make it difficult to predict message contents, (2 bis) even if it did, `Math.random()` isn't really unpredictable, (3) predicting message contents isn't bad in itself as the DB should be assumed to be public, (4) conflicts due to missing a nonce most certainly mean there's a bug.

Reply:
OK, I think I'm going to add it back in because of #1503 and it sounds like there's nothing wrong with it.

Reply:
Well, `Math.random()` was probably something that was wrong with it (if it needs to be cryptographically secure at least). I think that #1503 / #1486 probably point out to other issues.

> Messages can be re-created by 'chelonia/private/out/publishEvent', causing the hash of the successful message to be different, resulting in a hanging promise.

I'd say that the conditions that lead to a hanging promise seem to be the issue here, as not having a nonce isn't the only reason why this might happen. In other words, hanging promises should be handled for what they are, but that's a separate issue.

Other than that, I don't think a nonce is actively bad here, just unnecessary. The danger is relying on it being unique, which can't be guaranteed (realistically).

Reply:
Just an FYI this `nonce` was recently added back in a recent PR. In the future, if we want to get rid of it, it will require changing how chat pending messages work by setting the `id` in the data when creating a chat message. Each chat message creation would need to generate its own unique ID, and then it wouldn't need to rely on the GIMessage `.id()` API